homework/zad9_opencv/pattern_match.cpp

```diff
@@ -0,0 +1,87 @@
+#include "opencv2/imgcodecs.hpp"
+#include "opencv2/highgui.hpp"
+#include "opencv2/imgproc.hpp"
+#include <iostream>
+#include <mutex>
+#include <future>
+#include <thread>
+
+int main( int argc, char** argv )
+{
+    if (argc < 3)
+    {
+        std::cout << "Not enough parameters" << std::endl;
+        std::cout << "Usage:\n" << argv[0] << " <source_name> <pattern1_name> [<pattern2_name> ...]" << std::endl;
+        return -1;
+    }
+
+    const int patternsCnt = argc - 2;
+
+    cv::Mat imgSource;
+    cv::Mat imgPatterns[patternsCnt];
+
+    // load images
+    imgSource = cv::imread( argv[1], cv::IMREAD_COLOR );
+    for ( int i = 0; i < patternsCnt; ++i)
+    {
+        imgPatterns[i] = cv::imread( argv[i + 2], cv::IMREAD_COLOR );
+    }
+
+    // create windows
+    const char* image_window = "Source Image";
+    cv::namedWindow( image_window, cv::WINDOW_AUTOSIZE );
+
+    // copy source image
+    cv::Mat img_display;
+    imgSource.copyTo( img_display );
+
+    std::vector<std::future<int>> futures(patternsCnt);
+    std::mutex criticalSectionMutex;
+
+    for( int i = 0; i < patternsCnt; ++i )
+    {
+        auto matchFunction = [&criticalSectionMutex](const cv::Mat& source, const cv::Mat& pattern, cv::Mat& display)
+        {
+            cv::Mat result;
+
+            // Create the result matrix
+            const int result_cols =  source.cols - pattern.cols + 1;
+            const int result_rows = source.rows - pattern.rows + 1;
+
+            result.create( result_rows, result_cols, CV_32FC1 );
+
+            // match pattern and normalize
+            cv::matchTemplate(source, pattern, result, cv::TM_SQDIFF);
+            cv::normalize(result, result, 0, 1, cv::NORM_MINMAX, -1, cv::Mat());
```

Comment:
To jest część robiąca właściwe wykrywanie wzorca. Reszta poniżej to jest szukanie gdzie ten wzorzec jest, aby wyświetlić to dopasowanie i to może być rezultatem zwracanym przez obecne zadanie. Wyświetlanie dopasowań może być robione przez zupełnie oddzielny wątek, który będzie służył tylko do wyświetlania.

Reply:
ok